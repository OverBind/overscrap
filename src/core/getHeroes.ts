import puppeteer from 'puppeteer';
import { Hero } from '../types';
import { HeroID, HeroStat } from '../enums';

export async function getHeroes(page: puppeteer.Page): Promise<Hero[]> {
  return page.evaluate(
    (HeroID, HeroStat) => {
      const list: Hero[] = [];
      const items = document.querySelectorAll(
        'div#competitive select.dropdown-select-element option'
      );

      items.forEach((item) => {
        const id: string = item.getAttribute('value') || '';
        const name = item.innerHTML;
        const isHero = Object.values(HeroID).find(
          (heroID) => heroID === id && id !== HeroID.ALL_HEROES
        );
        if (isHero) {
          const hero = {
            id,
            name,
            src: `https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/${id}.png`,
            info: Object.values(HeroStat).map(
              (dataId) => {
                const getHeroStats: HTMLInputElement | null =
                  document.querySelector(
                    `div#competitive div[data-category-id = '${id}']`
                  );
                if (getHeroStats) {
                  const data = Array.from(
                    getHeroStats.querySelectorAll(
                      `tr[data-stat-id = '${dataId}'] td`
                    ),
                    (col) => col.innerHTML
                  );
                  return {
                    name: data[0],
                    value: data[1]
                  };
                }
                return {
                  name: '',
                  value: ''
                };
              }
            )
          };
          list.push(hero);
        }
      });
      return list;
    },
    HeroID,
    HeroStat
  );
}
