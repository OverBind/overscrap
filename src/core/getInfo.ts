import puppeteer from 'puppeteer';
import { GameStat } from '../enums';
import { InfoField } from '../types';

export async function getInfo(page: puppeteer.Page): Promise<InfoField[]> {
  return page.evaluate((GameStat) => {
    return Object.values(GameStat).map((dataId) => {
      const row = Array.from(
        document.querySelectorAll(`tr[data-stat-id = '${dataId}']`)
      );
      const data = Array.from(row[1].children, (td) => td.innerHTML);
      return {
        name: data[0],
        value: data[1]
      };
    });
  }, GameStat);
}
