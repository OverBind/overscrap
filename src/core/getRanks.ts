import puppeteer from 'puppeteer';
import { Rank } from '../types';

export async function getRanks(page: puppeteer.Page): Promise<Rank[]> {
  return page.evaluate(() => {
    return Array.from(
      document.querySelectorAll(
        'div.masthead-player div.competitive-rank-role'
      ),
      (rank) => ({
        src: rank.querySelector<HTMLInputElement>(
          'img.competitive-rank-tier-icon'
        )?.src,
        role: rank
          .querySelector<HTMLInputElement>('span.ow-Tooltip')
          ?.innerText.replace(/ .*/, ''),
        sr: rank.querySelector<HTMLInputElement>('div.competitive-rank-level')
          ?.innerText
      })
    );
  });
}
