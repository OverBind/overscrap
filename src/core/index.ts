export { getHeroes } from './getHeroes';
export { getInfo } from './getInfo';
export { getRanks } from './getRanks';