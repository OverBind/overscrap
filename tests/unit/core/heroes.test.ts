/**
 * @jest-environment jsdom
 */
import { setupTest } from '../../helper/setupTest';
import { getHeroes } from '../../../src/core/getHeroes';

describe('Get Heroes', () => {
  test('Should get heroes data', async () => {
    const page = await setupTest();

    const data = await getHeroes(page);

    expect(data).toBeTruthy();
    expect(data).toMatchSnapshot();
  });
});
