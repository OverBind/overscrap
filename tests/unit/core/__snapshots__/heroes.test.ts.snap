// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Heroes Should get heroes data 1`] = `
Array [
  Object {
    "id": "0x02E000000000007A",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "4",
      },
      Object {
        "name": "Games Played",
        "value": "10",
      },
      Object {
        "name": "Games Won",
        "value": "5",
      },
      Object {
        "name": "Time Played",
        "value": "01:35:53",
      },
      Object {
        "name": "Win Percentage",
        "value": "54%",
      },
    ],
    "name": "D.Va",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E000000000007A.png",
  },
  Object {
    "id": "0x02E000000000013E",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "2",
      },
      Object {
        "name": "Games Played",
        "value": "6",
      },
      Object {
        "name": "Games Won",
        "value": "4",
      },
      Object {
        "name": "Time Played",
        "value": "01:15:58",
      },
      Object {
        "name": "Win Percentage",
        "value": "62%",
      },
    ],
    "name": "Orisa",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E000000000013E.png",
  },
  Object {
    "id": "0x02E0000000000007",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "10",
      },
      Object {
        "name": "Games Played",
        "value": "20",
      },
      Object {
        "name": "Games Won",
        "value": "10",
      },
      Object {
        "name": "Time Played",
        "value": "03:40:57",
      },
      Object {
        "name": "Win Percentage",
        "value": "51%",
      },
    ],
    "name": "Reinhardt",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E0000000000007.png",
  },
  Object {
    "id": "0x02E0000000000040",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "6",
      },
      Object {
        "name": "Games Played",
        "value": "15",
      },
      Object {
        "name": "Games Won",
        "value": "8",
      },
      Object {
        "name": "Time Played",
        "value": "02:26:14",
      },
      Object {
        "name": "Win Percentage",
        "value": "57%",
      },
    ],
    "name": "Roadhog",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E0000000000040.png",
  },
  Object {
    "id": "0x02E000000000023B",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "8",
      },
      Object {
        "name": "Games Played",
        "value": "14",
      },
      Object {
        "name": "Games Won",
        "value": "6",
      },
      Object {
        "name": "Time Played",
        "value": "02:52:11",
      },
      Object {
        "name": "Win Percentage",
        "value": "41%",
      },
    ],
    "name": "Sigma",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E000000000023B.png",
  },
  Object {
    "id": "0x02E0000000000009",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "5",
      },
      Object {
        "name": "Games Played",
        "value": "13",
      },
      Object {
        "name": "Games Won",
        "value": "8",
      },
      Object {
        "name": "Time Played",
        "value": "02:30:04",
      },
      Object {
        "name": "Win Percentage",
        "value": "62%",
      },
    ],
    "name": "Winston",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E0000000000009.png",
  },
  Object {
    "id": "0x02E00000000001CA",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "5",
      },
      Object {
        "name": "Games Played",
        "value": "9",
      },
      Object {
        "name": "Games Won",
        "value": "3",
      },
      Object {
        "name": "Time Played",
        "value": "01:31:47",
      },
      Object {
        "name": "Win Percentage",
        "value": "38%",
      },
    ],
    "name": "Wrecking Ball",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E00000000001CA.png",
  },
  Object {
    "id": "0x02E0000000000068",
    "info": Array [
      Object {
        "name": "Games Lost",
        "value": "2",
      },
      Object {
        "name": "Games Played",
        "value": "3",
      },
      Object {
        "name": "Games Won",
        "value": "1",
      },
      Object {
        "name": "Time Played",
        "value": "32:15",
      },
      Object {
        "name": "Win Percentage",
        "value": "28%",
      },
    ],
    "name": "Zarya",
    "src": "https://d1u1mce87gyfbn.cloudfront.net/game/heroes/small/0x02E0000000000068.png",
  },
]
`;
